# ADM - HW 3

In [1]:
import functions 

from bs4 import BeautifulSoup
import requests
import os 
import time
import pandas as pd 
import csv 
import nltk
import calendar
from nltk.corpus import stopwords
import numpy as np
from collections import Counter
from functools import reduce
import pickle 
from sklearn.feature_extraction.text import TfidfVectorizer, TfidfTransformer
from sklearn.metrics.pairwise import cosine_similarity
import heapq
from collections import Counter
import re


# 1. Data Collection

## 1.1 Get the list of master's degree courses

In [ ]:
# se to true to install bs4 & selenium
if False: 
    pip install bs4;
    pip install selenium

First of all, we get a list of all the master's degree courses we need.

In [2]:
def extract_masters(this_url):
  result_url = requests.get(this_url)
  result_soup = BeautifulSoup(result_url.text)
  result_links = result_soup.find_all('a', {'class':'courseLink'})
  result_list=[]
  for item in result_links:
    result_list.append((item['href'], item.text))

  return result_list

We notice that the url of th various pages is *https://www.findamasters.com/masters-degrees/msc-degrees/?PG=* followed by the number of the page in the MSc courses catalogue we're in. 

In [3]:
links = []
for page in range(1, 401):
    url = 'https://www.findamasters.com/masters-degrees/msc-degrees/?PG='+str(page)
    links += extract_masters(url)
    print(page, len(links))

1 15
2 30
3 45
4 60
5 75
6 90
7 105
8 120
9 135
10 150
11 165
12 180
13 195
14 210
15 225
16 240
17 255
18 270
19 285
20 300
21 315
22 330
23 345
24 360
25 375
26 390
27 405
28 420
29 435
30 450
31 465
32 480
33 495
34 510
35 525
36 540
37 555
38 570
39 585
40 600
41 615
42 630
43 645
44 660
45 675
46 690
47 705
48 720
49 735
50 750
51 765
52 780
53 795
54 810
55 825
56 840
57 855
58 870
59 885
60 900
61 915
62 930
63 945
64 960
65 975
66 990
67 1005
68 1020
69 1035
70 1050
71 1065
72 1080
73 1095
74 1110
75 1125
76 1140
77 1155
78 1170
79 1185
80 1200
81 1215
82 1230
83 1245
84 1260
85 1275
86 1290
87 1305
88 1320
89 1335
90 1350
91 1365
92 1380
93 1395
94 1410
95 1425
96 1440
97 1455
98 1470
99 1485
100 1500
101 1515
102 1530
103 1545
104 1560
105 1575
106 1590
107 1605
108 1620
109 1635
110 1650
111 1665
112 1680
113 1695
114 1710
115 1725
116 1740
117 1755
118 1770
119 1785
120 1800
121 1815
122 1830
123 1845
124 1860
125 1875
126 1890
127 1905
128 1920
129 1935
130 1950
131 1965
1

We want to verify we retrieved all of the needed urls, so we check for the length of the list

In [5]:
print(len(links))

6000


It's the expected length, so we can create and write the required *.txt* file.   

In [4]:
file = open('msc_links.txt','w')
for link in links:
	file.write(link[0]+'\n')

file.close()

Now, if we look at the file it's written in the correct way (:

## 1.2 Crawl master's degree pages 

In [62]:
def download_html(url, dest_folder):
    #path is the local path we're working in 
    if not os.path.exists(dest_folder):  # create folder if it does not exist
        os.makedirs(dest_folder) 
    
    filename = url.replace('https://www.findamasters.com//masters-degrees/course/', '').replace('/', '_')
    file_path = os.path.join(dest_folder, filename)
    url_req = requests.get(url)

    if url_req.ok:
        soup = BeautifulSoup(url_req.text, 'html.parser')
        with open(file_path, 'w') as file: 
            file.write(soup.prettify())
        
        file.close()
        

In [68]:
### SAVING HTML & ORGANIZING THEM IN FOLDERS ###
file = open('msc_links.txt', 'r')
lines_list = file.readlines()
file.close()

In [73]:
for i in range(1, 401):
    folder_i = 'page_' + str(i)
        
    #selecting the lines corresponding to the i-th page
    start = (i-1)*15
    stop =  i*15
    needed_lines = lines_list[start : stop]

    #saving the corresponding html in "page_i" folder
    for line in needed_lines:
        url  = 'https://www.findamasters.com/' + str(line)[:-1]
        download_html(str(url), folder_i)
        time.sleep(2) 
    

## 1.3 Parse downloaded pages

Creating a DataSet starting from the scrapped *.tsv* files. 

In [2]:
msc_degrees = pd.DataFrame(columns=['courseName', 'universityName', 'facultyName', 'isItFullTime', 'description', 'startDate', 'fees', 'modality', 'duration', 'city', 'country', 'administration', 'url'])
for i in range(1, 6001): 
    course_file = csv.reader(open(os.path.join('courses_tsv', 'course_'+str(i))+'.tsv', 'r'), delimiter='\t')
    line = pd.DataFrame(course_file, columns=['courseName', 'universityName', 'facultyName', 'isItFullTime', 'description', 'startDate', 'fees', 'modality', 'duration', 'city', 'country', 'administration', 'url'], index = [i])
    msc_degrees = pd.concat([msc_degrees, line])


To see if we created the DataFrame in the correct way 

In [3]:
msc_degrees

courseName  \
1                 Advanced Physiotherapy Practice - MSc   
2     Advanced Master in Innovation & Strategic Mana...   
3     Accounting, Financial Management and Digital B...   
4                             Analytical Toxicology MSc   
5                          Accounting and Finance - MSc   
...                                                 ...   
5996  Masters of Science in Business, Supply Chain A...   
5997    Material Culture & Artefact Studies - MSc/PgDip   
5998                       Master's of Computer Science   
5999         Master's of Financial Technology (Fintech)   
6000                                 Masters of Finance   

                      universityName                           facultyName  \
1      Glasgow Caledonian University    School of Health and Life Sciences   
2             Solvay Brussels School              Economics and Management   
3              University of Reading                Henley Business School   
4              King’s College London   Faculty of Life Sciences & Medicine   
5                University of Leeds      Leeds University Business School   
...                              ...                                   ...   
5996         Oregon State University                    School of Business   
5997           University of Glasgow          College of Arts & Humanities   
5998        Harbour.Space University                    Masters Programmes   
5999        Harbour.Space University                    Masters Programmes   
6000         University of Hong Kong     Faculty of Business and Economics   

     isItFullTime                                        description  \
1       Full time   Progress your career as a physiotherapist wit...   
2       Full time   Programme overview   The Advanced Master in I...   
3       Full time   Embark on a professional accounting career wi...   
4       Full time   The Analytical Toxicology MSc is a unique stu...   
5       Full time   Businesses and governments rely on sound fina...   
...           ...                                                ...   
5996    Full time   Master of Science in Business (MSB)   Our Mas...   
5997    Full time   Material culture and artefact studies combine...   
5998    Full time   Harbour.Space’s Master's of Computer Science ...   
5999    Full time   Harbour.Space's FinTech Master programme     ...   
6000    Full time   The HKU Business School Master of Finance (MF...   

                startDate                                               fees  \
1      January, September   Please see the university website for further...   
2               September                                           18.000 €   
3               September   Please see the university website for further...   
4              See Course   Please see the university website for further...   
5               September   UK: £18,000 (Total)  International: £34,750 (...   
...                   ...                                                ...   
5996           See Course   Please see the university website for further...   
5997            September   Please see the university website for further...   
5998   September, January                                       €29,900/year   
5999   September, January                                       €29,900/year   
6000               August   Please see the university website for further...   

                                      modality  \
1                                          MSc   
2      MA MSc Other Pre-Masters Masters Module   
3                                          MSc   
4                                          MSc   
5                                          MSc   
...                                        ...   
5996                                       MSc   
5997                                 MSc PGDip   
5998                                       MSc   
5999                                   MBA MSc   
6000                   

# 2. Search Engine

## 2.0 Preprocessing

### 2.0.0  Preprocessing the text

We are required to preprocess all the informations retrieved for the courses, but since loosing punctuation in the url will lead us to loose its functionality, we won't preprocess that column (also, the info contained in the url aren't textual info). Moreover, we won't preprocess the fees column since there's a particoular preprocessing required for that column.

- 'description': removing punctuation and stopwords, stemming
- 'startDate': eliminating everithing that is not a month     (oss: calendar.month_name contains an empty name as first element)
- 'courseName': eliminating 'Msc' and similar: lo faccio dopo 


In [3]:
# set to True to download 
if False:
    nltk.download('punkt') 
    nltk.download('stopwords')

In [10]:
#'descritpion' column 
stopw = stopwords.words('english')
msc_degrees['description_clean'] = msc_degrees.description.apply(lambda row: nltk.word_tokenize(row)).apply(lambda row: ([nltk.PorterStemmer().stem(word) for word in row if (word.isalnum() and (not word in stopw))])) 

#'startDate' column
msc_degrees['startDate_clean']= msc_degrees.startDate.apply(lambda row: nltk.word_tokenize(row)).apply(lambda row: ' '.join([month for month in row if month in calendar.month_name[1:]]))

#'courseName' column: TBD


In [11]:
msc_degrees

courseName  \
1                 Advanced Physiotherapy Practice - MSc   
2     Advanced Master in Innovation & Strategic Mana...   
3     Accounting, Financial Management and Digital B...   
4                             Analytical Toxicology MSc   
5                          Accounting and Finance - MSc   
...                                                 ...   
5996  Masters of Science in Business, Supply Chain A...   
5997    Material Culture & Artefact Studies - MSc/PgDip   
5998                       Master's of Computer Science   
5999         Master's of Financial Technology (Fintech)   
6000                                 Masters of Finance   

                      universityName                           facultyName  \
1      Glasgow Caledonian University    School of Health and Life Sciences   
2             Solvay Brussels School              Economics and Management   
3              University of Reading                Henley Business School   
4              King’s College London   Faculty of Life Sciences & Medicine   
5                University of Leeds      Leeds University Business School   
...                              ...                                   ...   
5996         Oregon State University                    School of Business   
5997           University of Glasgow          College of Arts & Humanities   
5998        Harbour.Space University                    Masters Programmes   
5999        Harbour.Space University                    Masters Programmes   
6000         University of Hong Kong     Faculty of Business and Economics   

     isItFullTime                                        description  \
1       Full time   Progress your career as a physiotherapist wit...   
2       Full time   Programme overview   The Advanced Master in I...   
3       Full time   Embark on a professional accounting career wi...   
4       Full time   The Analytical Toxicology MSc is a unique stu...   
5       Full time   Businesses and governments rely on sound fina...   
...           ...                                                ...   
5996    Full time   Master of Science in Business (MSB)   Our Mas...   
5997    Full time   Material culture and artefact studies combine...   
5998    Full time   Harbour.Space’s Master's of Computer Science ...   
5999    Full time   Harbour.Space's FinTech Master programme     ...   
6000    Full time   The HKU Business School Master of Finance (MF...   

                startDate                                               fees  \
1      January, September   Please see the university website for further...   
2               September                                           18.000 €   
3               September   Please see the university website for further...   
4              See Course   Please see the university website for further...   
5               September   UK: £18,000 (Total)  International: £34,750 (...   
...                   ...                                                ...   
5996           See Course   Please see the university website for further...   
5997            September   Please see the university website for further...   
5998   September, January                                       €29,900/year   
5999   September, January                                       €29,900/year   
6000               August   Please see the university website for further...   

                                      modality  \
1                                          MSc   
2      MA MSc Other Pre-Masters Masters Module   
3                                          MSc   
4                                          MSc   
5                                          MSc   
...                                        ...   
5996                                       MSc   
5997                                 MSc PGDip   
5998                                       MSc   
5999                                   MBA MSc   
6000                   

### 2.0.1 - Preprocessing the fees columns

First of all, we notice that there are a lot of courses for which we are addressed to the University's site to see fees' informations, and for these courses, th 'fees' columns says: *"Please see the university website for further information on fees for this course."*. For this reason, we replace such stirng with an empty one under the *'fees_clean'* column in order to do superflous operations.

In [3]:
msc_degrees['fees_clean'] = msc_degrees.fees.apply(lambda row: row.strip().replace('Please see the university website for further information on fees for this course.', ''))

In [13]:
len(msc_degrees[msc_degrees.fees_clean == ''])

4009

We notice that we now have 4009 empty lines for the fees values, which makes our operation actually useful.  
Now let's get started eith the rest of the pre-proessing of the fees column.

problem : between the ones withouth matches, we have a lot of rows with structure like 
"UK Fees: 2022/23 fees TBC*;2021/22 fees - 10400   International Fees: 2022/23 fees TBC*;2021/22 fees - 17900".


For these lines, the country is always UK, as we suspected, so we assume that the currency is £, even if not explicitly written. 
For these reasons, for these lines we procede in the following way:
1. match 'UK FEES' at the beginnning of the line
2. search for all the numeric values, clean them by removing '.' and ','
3. save numerical values + £ in the 'fees_clean' column 


In [4]:
#cleaning fees column

# defining regex atterns to look for:
#pattern1: currency, numeric value
#pattern2: numeric value, currency
#pteernUK: matches its content at the beginning of the string (see explaination above in markdown)
pattern1 = re.compile(u'([$¢£¤¥֏؋৲৳৻૱௹฿៛\u20a0-\u20bd\uff04\uffe0\uffe1\uffe5\uffe6]|'r'USD|EUR|GBP|JPY|INR|AUD|CAD|HK|euro|euros|US$|ISK|RMB|SEK|Euro|Euros|CHF|Eur)'r'( {0,1}[0-9]+[.,]{0,1}[0-9]+)+')
pattern2 = re.compile(r'([0-9]+[.,]{0,1}[0-9]+)+ {0,1}'u'([$¢£¤¥֏؋৲৳৻૱௹฿៛\u20a0-\u20bd\uff04\uffe0\uffe1\uffe5\uffe6]|'r'USD|EUR|GBP|JPY|INR|AUD|CAD|HK|euro|euros|US$|ISK|RMB|SEK|Euro|Euros|CHF|Eur)')
pattern_UKfees = r'^UK Fees:'


msc_degrees['fees_clean'] = msc_degrees.fees_clean.apply(lambda row: re.findall(pattern1, row) + [el[::-1] for el in re.findall(pattern2, row)] +                                          #all matches for pattern1 & pattern 2 in a list of tuples + reversing pattern2 matches
                                                         [ ('£', numeric_value.strip()) for numeric_value in re.findall(r'[ *]{1}[0-9]+ ', row + ' ') if re.search(pattern_UKfees, row.lstrip()) ] #'UK Fees:' lines: matching and retrieving needed numeric values + storing them in appropriate way
                                                         ).apply(lambda row: [(fees[0], fees[1].replace('.', '').replace(',','').replace('*','')) for fees in row])                        #cleaning numeric values
                                                                 

Prima dei cambiare le tutte le currency in simboli con le sigle !!!!!!!!!!!


In [6]:
ISO_currency_dict = {
    '$' : 'USD',
    '£' : 'GBP',
    '¥': 'JPY',
    '֏': 'AMD',
    '؋': 'AFN',
    '৲': 'BDT',
    '৳': 'BDT',
    '৻': 'VES',
    '૱': 'KHR',
    '௹': 'LKR',
    '฿': 'THB',
    '៛': 'KHR',
    'euro' : 'EUR',
    'euros' : 'EUR',
    'Euro' : 'EUR',
    'Euros' : 'EUR',
    'Eur' : 'EUR',
    'US$' : 'USD',
    'USD': 'USD',
    'EUR': 'EUR',
    'GBP': 'GBP',
    'JPY': 'JPY',
    'INR': 'INR',
    'AUD': 'AUD',
    'CAD': 'CAD',
    'HK': 'HKD',
    'ISK': 'ISK',
    'RMB': 'CNY',
    'SEK': 'SEK',
    'CHF': 'CHF',
    
    #'\u20a0-\u20bd': below
    '₡': 'CRC', #Costa Rican Colón
    '₤': 'ITL', # Italian Lira
    '₥': 'DEM', # German Mark
    '₦': 'NGN', # Nigerian Naira
    '₧': 'ESP', # Spanish Peseta
    '₨': 'INR', # Indian Rupee
    '₩': 'KRW', # South Korean Won
    '₪': 'ILS', # Israeli New Shekel
    '₫': 'VND', # Vietnamese Đồng
    '€': 'EUR', # Euro
    '₭': 'LAK', # Lao Kip
    '₮': 'MNT', # Mongolian Tugrik
    '₯': 'GRD', # Greek Drachma
    '₱': 'PHP', # Philippine Peso
    '₲': 'PYG', # Paraguayan Guarani
    '₴': 'UAH', # Ukrainian Hryvnia
    '₵': 'GHS', # Ghanaian Cedi    
}

In [16]:
def currency_converter(money:tuple, currency_to:str, currency_dict: dict):
    currency_from, numeric_value = money

    #making sure the "output currency" is a abbreviation and not a symbol
    if (not currency_to in currency_dict) and (not currency_from in currency_dict):
        print(f'Unable to convert this currency')
        return
    
    #making sure the "output currency" is a abbreviation and not a symbol
    currency_to = currency_dict[currency_to]
    currency_from = currency_dict[currency_from]
    response = requests.get('https://open.er-api.com/v6/latest/' + currency_from)
    time.sleep(1)

    if response.ok:
        data = response.json()
        if data['result'] != 'error':
            change_rates = data['rates']

        return int(numeric_value)*change_rates[currency_to]


In [32]:
def get_max_currency(lst):
    if lst ==[]:
        return 
    else: 
        return max(lst)

In [36]:
msc_degrees['fees_EUR'] = msc_degrees.fees_clean.apply(lambda row: [functions.currency_converter(el, 'EUR', ISO_currency_dict) for el in row])
msc_degrees['fees_EUR'] = msc_degrees.fees_EUR.apply(functions.get_max_currency)

## 2.1 Conjunctive query 

### 2.1.1 Create your index!

#### Creating *vocabulary* file.     
To do so, we first create a list, *vocabulary_list*, containing all the words in the '*description_clean*' column of the dataset, then we set each word's *term_id* as its implicit index in this list. To do so, zip the previous list with another explicitly containg the indeces we choose.
''', converted in string type (this way we don't have any problem when writing the *vocabulary.tsv* file).'''

In [5]:
# set to True to re-write the vocabulary file
if False:
    # creating vocabulary lists as a tool to write the vocabulary file 
    vocabulary_list = sorted(list(Counter(reduce(lambda x, y: x+y, msc_degrees.description_clean.values)).keys()))
    vocabulary_idx = dict(zip(vocabulary_list, list(range( len(vocabulary_list)))))
    
    # save dictionary to vocabulary.pkl file
    with open('vocabulary.pkl', 'wb') as v_file:
        pickle.dump(vocabulary_idx, v_file)

    v_file.close()

#### Inverted index creation

In [8]:
# set to True to re-write the inverted index file
if False: 
    # Read dictionary pkl file
    with open('vocabulary.pkl', 'rb') as v_file:
        vocabulary_dict = pickle.load(v_file)

    #function to get the documents in which term appears
    get_documents_list = lambda word:  list(msc_degrees.loc[msc_degrees.description_clean.apply(lambda row: word in row )].index)

    #initializing inverted_index_dict
    inverted_index = dict()

    #updating the inverted_index dictionary
    for key in vocabulary_dict:
        #saving needed values in order to call the funcion only once per term 
        doc_list =  get_documents_list(key) 

        #updating dict
        inverted_index.update({vocabulary_dict[key] : doc_list})

    # save dictionary to inverted_index.pkl file
    with open('inverted_index.pkl', 'wb') as invidx_file:
        pickle.dump(inverted_index, invidx_file)

    invidx_file.close()


### 2.1.2 Executing the query

First of all, we need to load the files for the query execution

In [4]:
# Read dictionary pkl file
with open('vocabulary.pkl', 'rb') as v_file:
    vocabulary_dict = pickle.load(v_file)

v_file.close()

#Read inverted_index pkl file
with open('inverted_index.pkl', 'rb') as invidx_file:
    inverted_index = pickle.load(invidx_file)

invidx_file.close()

- Now we have to preprocess the query in the same way we preprocessed the 'description' column to get a match: We define a function that does that, then we apply it to the query
- then we get the index of the documents_ids in which the words appear 
- return list of dovìcuments_ids

In [13]:
#getting the query from input and stripping it
query = input('Type to search:').strip()

#preprocessing the query
query = functions.preprocess_query(query)

#searching documents
doc_sat_query = functions.get_documents_conjunctive_query(query, vocabulary_dict, inverted_index)

#Results visualizatioin
if doc_sat_query == []:
    print("This query didn't produce any result")
else: 
    print('Results:')

results_columns = ['courseName', 'universityName', 'description', 'url']
msc_degrees.loc[doc_sat_query, results_columns]

Now we want to visualize the results given from the query

In [14]:
if doc_sat_query == []:
    print("This query didn't produce any result")
else: 
    print('Results:')

results_columns = ['courseName', 'universityName', 'description', 'url']
msc_degrees.loc[doc_sat_query, results_columns]

Results:


courseName  \
2048                           Clinical Optometry - MSc   
4098  Global Meetings and Events Management MSc / PGDip   
4                             Analytical Toxicology MSc   
5                          Accounting and Finance - MSc   
8                                        Addictions MSc   
...                                                 ...   
1991                    Clinical Embryology - MSc/PGDip   
1999                          Clinical Geriatrics - MSc   
2021                     Clinical Neuropsychology - MSc   
2029                   Clinical Ophthalmic Practice MSc   
2047                            Clinical Pharmacy - MSc   

                      universityName  \
2048              Cardiff University   
4098   University College Birmingham   
4              King’s College London   
5                University of Leeds   
8              King’s College London   
...                              ...   
1991             University of Leeds   
1999              Cardiff University   
2021           University of Bristol   
2029       University College London   
2047        University of Sunderland   

                                            description  \
2048   Why study this course    The aim of this prog...   
4098   Become part of an events industry worth an es...   
4      The Analytical Toxicology MSc is a unique stu...   
5      Businesses and governments rely on sound fina...   
8      Join us for an online session for prospective...   
...                                                 ...   
1991   The Leeds MSc in Clinical Embryology is a par...   
1999   Why study this course    The MSc Clinical Ger...   
2021   Professional programmes in Clinical Neuropsyc...   
2029   Register your interest in graduate study at U...   
2047   The Clinical Pharmacy MSc has been designed t...   

                                                    url  
2048   https://www.findamasters.com/masters-degrees/...  
4098   https://www.findamasters.com/masters-degrees/...  
4      https://www.findamasters.com/masters-degrees/...  
5      https://www.findamasters.com/masters-degrees/...  
8      https://www.findamasters.com/masters-degrees/...  
...                                                 ...  
1991   https://www.findamasters.com/masters-degrees/...  
1999   https://www.findamasters.com/masters-degrees/...  
2021   https://www.findamasters.com/masters-degrees/...  
2029   https://www.findamasters.com/masters-degrees/...  
2047   https://www.findamasters.com/masters-degrees/...  

[480 rows x 4 columns]

Checking if we have the TAs' restults

In [15]:
res = msc_degrees.loc[doc_sat_query, results_columns]
names = ['Analytical Toxicology MSc', 'Addictions MSc', 'Accounting and Finance - MSc', 'Allergy - MSc/PGDip/PGCert']
res[res['courseName'].isin(names)]

courseName            universityName  \
4        Analytical Toxicology MSc     King’s College London   
5     Accounting and Finance - MSc       University of Leeds   
8                   Addictions MSc     King’s College London   
528   Accounting and Finance - MSc      University of Sussex   
1012    Allergy - MSc/PGDip/PGCert   Imperial College London   

                                            description  \
4      The Analytical Toxicology MSc is a unique stu...   
5      Businesses and governments rely on sound fina...   
8      Join us for an online session for prospective...   
528    On this MSc you’ll advance your accounting an...   
1012   Allergy is an increasing global health proble...   

                                                    url  
4      https://www.findamasters.com/masters-degrees/...  
5      https://www.findamasters.com/masters-degrees/...  
8      https://www.findamasters.com/masters-degrees/...  
528    https://www.findamasters.com/masters-degrees/...  
1012   https://www.findamasters.com/masters-degrees/...

## 2.2 Conjunctive query & Ranking score

### 2.2.1 Inverted index - second version

To create the inverted_index_tfidf dictionary, we observe that the ids we gave to the terms in the vocabulary are the indeces in the previous vocabulary list sorted by alphabetial order. When computing the tfidf score, by looking at a dataframe computed by the matrix, by changing the names of the columns, we saw that the terms wre stored in alhabetical order, leading us to be able to use the tfidf matrix indeces to retrieve the tfidf score for every couple term-document we were interested in. + we used the old inverted_index dictionary. 

In [55]:
if False:
    # calculating tfidf 
    tfidf = TfidfVectorizer(input='content', lowercase=False, tokenizer=lambda text: text) 
    tfidf_matrix = tfidf.fit_transform(msc_degrees.description_clean)

    #Read inverted_index pkl file (needed to construct inverted_index with tfidf)
    with open('inverted_index.pkl', 'rb') as invidx_file:
        inverted_index = pickle.load(invidx_file)

    invidx_file.close()

    #creating inverted index (with tfidf) dictionary
    inverted_index_tfidf = dict()
    for term_id in inverted_index:
        inverted_index_tfidf.update( { term_id: dict([ (doc_id, tfidf_matrix[(doc_id -1 , term_id)] ) for doc_id in inverted_index[term_id] ]) } )

    #saving the dictionary as a pickle file
    with open('inverted_index_tfidf.pkl', 'wb') as tfidf_file:
            pickle.dump(inverted_index_tfidf, tfidf_file)

    tfidf_file.close()

### 2.2.2 Execute the query

in order to get the top-k documents regarding the cosine similarity with the query, we decided to: 
1. preprocessing the query
2. getting all the documents containing the query (note that if a document doesn't contain the query, their cosine similarity in the tfidf representation is zero);
3. compute the cosine similarity of these retrived documents with the query, than storing them in a heap structure
4. return the top-k documents as a list
5. show results in a dataframe 

let us observe that we should take into account the different lengths of documents, which throw off the tfidf even with the same number of occurencies of a word (numerical values differing only slightly should be treated in the same way)

In [5]:
#Loading tfidf inverted index pkl file
with open('inverted_index_tfidf.pkl', 'rb') as invidx_file:
    inverted_index_tfidf = pickle.load(invidx_file)

invidx_file.close()

In [8]:
#point 1
query = input('Type to search: ')
query = functions.preprocess_query(query)

#get search results
topkresults_idx_cossim = functions.search_engine_tfidf(query, inverted_index_tfidf, vocabulary_dict, 20 )



In [11]:
#showing results
if topkresults_idx_cossim == []:
    print("There are no results for this query")
else: 
    results_columns = ['courseName', 'universityName', 'description', 'url']
    res = msc_degrees.loc[[el[0] for el in topkresults_idx_cossim], results_columns ]
    res['cossim'] = [el[1][0] for el in topkresults_idx_cossim]

res

courseName  \
2292                                          Computing   
942        Advancing Practice Sensory Integration (MSc)   
2256                               Computer Science MSc   
237                                     MSc - Economics   
2362                      Construction Management - MSc   
2365          Construction Management (part time) - MSc   
2437       Countering Extremist Crime and Terrorism MSc   
516                                    Accounting - MSc   
4806             International Financial Management MSc   
4901  International Master of Science in Fire Safety...   
915                     Advanced Sustainable Design MSc   
3290                             Engineering Design MSc   
1389  Aviation Operations and Management MSc (full t...   
5867           Master of Science in Nutritional Science   
1801  Caring for Children and Young People with Comp...   
1945  Civil Engineering: Innovative Structural Mater...   
4098  Global Meetings and Events Management MSc / PGDip   
2073         Clinical Speech and Language Studies - MSc   
4186                     Health and Social Practice MSc   
120                   Executive MSc Strategic Marketing   

                            universityName  \
2292             University of East London   
942            Sheffield Hallam University   
2256             University of East London   
237                      Durham University   
2362   Xi’an Jiaotong-Liverpool University   
2365   Xi’an Jiaotong-Liverpool University   
2437             University College London   
516                      Bangor University   
4806               University of Groningen   
4901               University of Edinburgh   
915                University of Edinburgh   
3290        Buckinghamshire New University   
1389        Buckinghamshire New University   
5867     University of Southern California   
1801            Queen’s University Belfast   
1945                    University of Bath   
4098         University College Birmingham   
2073                Trinity College Dublin   
4186                Northumbria University   
120                  King’s College London   

                                            description  \
2292   Do you want to be at the forefront of solving...   
942    Develop your knowledge and understanding of t...   
2256   Do you want to be at the forefront of solving...   
237    Our MSc programmes in Economics will give you...   
2362   In fast-growing countries around the world, t...   
2365   In fast-growing countries around the world, t...   
2437   Register your interest in graduate study at U...   
516    This degree programme provides the opportunit...   
4806   How do you manage international financial act...   
4901   Programme description    The International Ma...   
915    Programme description   Advanced Sustainable ...   
3290   This MSc in Engineering Design will equip you...   
1389   Whether you’re a recent aviation graduate or ...   
5867   The Master of Science in Nutritional Science ...   
1801   Advances in medical care and technology now m...   
1945   Advance your knowledge of civil engineering a...   
4098   Become part of an events industry worth an es...   
2073   The aims of the course are to provide qualifi...   
4186   If you want to advance your knowledge, unders...   
120    Looking to develop your marketing strategy sk...   

                                                    url    cossim  
2292   https://www.findamasters.com/masters-degrees/...  0.996022  
942    https://www.findamasters.com/masters-degrees/...  0.996022  
2256   https://www.findamasters.com/masters-degrees/...  0.996022  
237    https://www.findamasters.com/masters-degrees/...  0.993725  
2362   https://www.findamasters.com/masters-degrees/...  0.993725  
2365   https://www.findamasters.com/masters-degrees/...  0.993725  
2437   https://www.findamasters.com/masters-degrees/...  0.993725  
516    https://www.findamasters.com/masters-degrees/..